### PR TITLE
Add impl From<Duration> for Time and impl From<Time> for Duration

### LINKED ISSUE #216: time::Duration
We're using Symphonia in librespot as you know, and I was looking to moving most if not all time related things to Durations for consistency and ease of use. Anyway I noticed a lack of Duration support in your time units and calculations. And I was wondering if that was on purpose or not and if you'd be open to something like this?:

```rust
use std::{fmt, time::Duration};

...

impl From<Duration> for Time {
    fn from(duration: Duration) -> Self {
        let seconds = duration.as_secs_f64();
        Time::new(seconds.trunc() as u64, seconds.fract())
    }
}

impl TimeBase {
...
    pub fn calc_time_as_duration(&self, ts: TimeStamp) -> Duration {
        let time = self.calc_time(ts);
        Duration::from_secs_f64(time.seconds as f64 + time.frac)
    }
...
}
```

This would allow us to pass a Duration almost directly when seeking and get a duration for our packet time. If it's something that you're not interested in that's fine to, as it's pretty trivial to convert between Time and Duration ourselves.

### PR DESCRIPTION
**Pull request by JasonLG1979**
Closes #216

### CODE REVIEW

File: symphonia-core/src/units.rs [hunk 1]
```diff
@@ -70,6 +70,10 @@ impl Time {
 
         Some(Time { seconds, frac })
     }
+
+    pub fn into_duration(&self) -> StdDuration {
+        StdDuration::from_secs_f64(self.seconds as f64 + self.frac)
```
Comment (pdeljanov) [inline symphonia-core/src/units.rs:75]: A `f64` can only represent an unsigned integer up-to 2^52 without losing precision, so in extreme cases where seconds is very large, this conversion will be lossy.

Using `Duration::new` and passing the `u64` seconds component directly will preserve the accuracy in those extreme cases. To implement this, the fractional part must be converted to nanoseconds.

```suggestion
        StdDuration::new(self.seconds, (1_000_000_000.0 * self.frac) as u32)
```

File: symphonia-core/src/units.rs [hunk 2]
```diff
@@ -118,6 +122,13 @@ impl From<f64> for Time {
     }
 }
 
+impl From<StdDuration> for Time {
+    fn from(duration: StdDuration) -> Self {
+        let seconds = duration.as_secs_f64();
+        Time::new(seconds.trunc() as u64, seconds.fract())
```
Comment (pdeljanov) [inline symphonia-core/src/units.rs:128]: Same idea about improving precision by removing the round-trip through a `f64`.

```suggestion
        Time::new(duration.as_secs(), f64::from(duration.subsec_nanos()) / 1_000_000_000.0)
```

**Review by pdeljanov — CHANGES_REQUESTED** [review]
Thanks for this. Can you give these suggestions a try. I believe it should improve the accuracy in extreme cases.

File: symphonia-core/src/units.rs [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 //! The `units` module provides definitions for common units.
 
-use std::fmt;
+use std::{fmt, time::Duration as StdDuration};
```
[the remark this answers is not in the document]
Comment (JasonLG1979) [reply symphonia-core/src/units.rs]: Done

File: symphonia-core/src/units.rs (hunk 2 above)
Comment (JasonLG1979) [reply symphonia-core/src/units.rs]: Done.

File: symphonia-core/src/units.rs (hunk 1 above)
Comment (JasonLG1979) [reply symphonia-core/src/units.rs]: Done.

**Comment by pdeljanov** [comment]
Sorry for the possible churn, I had a last second thought about this. Do you think it'd be more idiomatic if instead of `into_duration()` we add `impl From<Time> for std::time::Duration {}`? Then one would just do `some_time.into()`.

**Comment by JasonLG1979** [comment]
> Sorry for the possible churn, I had a last second thought about this. Do you think it'd be more idiomatic if instead of `into_duration()` we add `impl From<Time> for std::time::Duration {}`? Then one would just do `some_time.into()`.

It's all good. Whatever you think is best. I am by no means a judge on what is more idiomatic.

**Comment by JasonLG1979** [comment]
There is some symmetry in impl'ing From for both that I can appreciate though. But I'm not really sure if it makes it more idiomatic or not?

**Comment by JasonLG1979** [comment]
Done.

**Review by pdeljanov — APPROVED** [review]

**Comment by pdeljanov** [comment]
I'm no expert either, but I do believe this very likely is the most idiomatic way. So let's go with it. Thanks for the making the changes.

**Pull request merged**